### DISCUSSION TITLE
Discussion: Functions with many parameters

### DISCUSSION THREAD

**Discussion by willow385** [rust-unofficial/patterns#239] [From Issues]
Requesting a design pattern for functions that take lots of parameters. For example this way:

```Rust
fn foo(parameter_0: type, parameter_1: type, parameter_2: type, parameter_3: type, parameter_4: type, parameter_5: type, parameter_6: type) -> type {
    // function body
}
```

is a way that I think should be an anti-pattern because it's too wide. I prefer things to be less than 80 chars wide.

**Comment by twe4ked** [comment]
My first thought would be, would it make sense for the parameters to be part of a struct that you construct and pass in? It would depend on what the actual parameters represent of course.

Another option could potentially be the [builder pattern](https://github.com/rust-unofficial/patterns/blob/829f1ef061ff39702d5029bbc93f0d37ae55af33/patterns/builder.md). Something like this:

```rust
foo()
    .parameter_0(1)
    .parameter_1(2)
    .parameter_2(3)
    .parameter_3(4)
    .parameter_4(5)
    .parameter_5(6)
    .parameter_6(7)
    .build();
```

EDIT: See https://github.com/rust-unofficial/patterns/issues/64 for reasons against builder patterns for some use cases.

**Comment by burdges** [comment]
```
pub struct DoSomething<..> { ..paramaters.. }
impl<..> DoSomething<..> {
    pub fn go(self) { ... }
}
```
And choose the parameter names well, so that you can maximize field puns. 

If you're okay with nightly then [`#![feature(unboxed_closures)]`](https://doc.rust-lang.org/nightly/unstable-book/language-features/unboxed-closures.html) and [`#![feature(fn_traits)]`](https://doc.rust-lang.org/nightly/unstable-book/library-features/fn-traits.html) let you write `DoSomething { .. }()` but imho that's kinda a poor reason for going to nightly.

**Comment by longfellowone** [comment]
> My first thought would be, would it make sense for the parameters to be part of a struct that you construct and pass in? It would depend on what the actual parameters represent of course.
> 
> Another option could potentially be the [builder pattern](https://github.com/rust-unofficial/patterns/blob/829f1ef061ff39702d5029bbc93f0d37ae55af33/patterns/builder.md). Something like this:
> 
> ```rust
> foo()
>     .parameter_0(1)
>     .parameter_1(2)
>     .parameter_2(3)
>     .parameter_3(4)
>     .parameter_4(5)
>     .parameter_5(6)
>     .parameter_6(7)
>     .build();
> ```
> 
> EDIT: See #64 for reasons against builder patterns for some use cases.

And how would you deal with mandatory parameters?

**Comment by twe4ked** [comment]
> And how would you deal with mandatory parameters?

I don't think I'd use it in that case.

**Comment by pickfire** [comment]
@longfellowone Do you have a required flow of the parameters?

Then there could be multiple structs (like a state machine) to ensure them have each mandatory parameter. Like most of the http crates that we have, it is required to have a url before sending.

**Comment by marcoieni** [comment]
> And how would you deal with mandatory parameters?

```rust
foo("mandatory1", "mandatory2")
    .parameter_0(1)
    .parameter_1(2)
    .parameter_2(3)
    .parameter_3(4)
    .parameter_4(5)
    .parameter_5(6)
    .parameter_6(7)
    .build();
```

**Comment by marcoieni** [comment]
I am thinking about how to solve this issue.
Maybe we can have an anti-pattern called "Too Many Arguments In Functions"?
It could explain all the approaches proposed in this issue.

**Comment by longfellowone** [comment]
> > And how would you deal with mandatory parameters?
> 
> ```rust
> foo("mandatory1", "mandatory2")
>     .parameter_0(1)
>     .parameter_1(2)
>     .parameter_2(3)
>     .parameter_3(4)
>     .parameter_4(5)
>     .parameter_5(6)
>     .parameter_6(7)
>     .build();
> ```

What I have is a mathematical calculation that requires about 7 mandatory parameters that have no reasonable defaults. And 5 parameters that are optional

**Comment by marcoieni** [comment]
Put those 7 parameters in a struct and give the struct to foo

**Comment by willow385** [comment]
Here's another idea:

```rust
fn foo(
    parameter_0: type,
    parameter_1: type,
    parameter_2: type,
    parameter_3: type,
    parameter_4: type,
    parameter_5: type,
    parameter_6: type
) -> type {
    // function body
}
```

that's what I've been doing in the years since I opened this ~~issue~~ discussion :)

**Reply by oiwn** [reply]
rustfmt doing that automatically.

**Comment by simonsan** [comment]
Some thoughts:

1. First, I'm unsure if having that many arguments to a function could be considered a code smell in itself. As in: "maybe this function/method is doing too much?"
1. Second, an alternative to the builder pattern could be to use `struct`s here:

```rust
pub struct FooDescriptor {
    parameter_0: type,
    parameter_1: type,
    parameter_2: Option<type>,
    parameter_3: type,
    parameter_4: type,
    parameter_5: type,
    parameter_6: Option<type>,
}

impl Default for FooDescriptor {
  // default()
}

fn foo(FooDescriptor) -> type {
   // function body
}

fn main() {
  let _ = foo(FooDescriptor::default());
}
```

That would make it easy to use optional arguments, have the arguments in one place, and would make it possible to define default parameters.

Examples:
- https://docs.rs/wgpu/latest/wgpu/struct.RenderPipelineDescriptor.html
- https://github.com/agausmann/global-clock/blob/cc9af2e255e3d5aa46ba05047efa079b6117ef82/src/globe.rs#L94-L303
